partitions.go

```diff
@@ -185,21 +185,34 @@ func (p *partitions) getAndCacheActivePartition(rootChecker func(StatCommander,
 	const devDir string = "/dev"
 
 	mountedDevices, err := getMountedDevices(devDir)
-	if err != nil {
-		return "", err
+	if err == nil {
+		activePartition, err := getRootFromMountedDevices(p, rootChecker, mountedDevices, rootDevice)
+		if err == nil {
+			if checkBootEnvAndRootPartitionMatch(bootEnvBootPart, activePartition) {
+				p.active = activePartition
+				log.Debug("Setting active partition: ", activePartition)
+				return p.active, nil
+			}
+
+			// If we reach this point, we have not been able to find a match
+			// based on mounted device.
+			//
+			// Fall-back to configuration and environment only!
+		}
```

Comment:
I'm a little bit confused about the use case and how this solves the problem. You have this in your mount output:

```
/dev/root on /rom type squashfs (ro,relatime)
```

However, neither `getMountedDevices()`[1] nor `getRootFromMountedDevices()` use this output, only `getRootCandidateFromMount()` does. `getMountedDevices()` gets the device files on the system, and then try to match this against the major and minor number of the `/` descriptor. Does squashfs not have a device file to match against?

If so, I think we should turn the matching against `p.rootfsPartA` and `p.rootfsPartB` into a special case rather than a fallback. IOW I think this code path should be taken if and only if `getRootFromMountedDevices()` returns error. Feels safer. :-) Here is my suggestion:

```
diff --git a/partitions.go b/partitions.go
index db67a30..5aba1d2 100644
--- a/partitions.go
+++ b/partitions.go
@@ -191,6 +191,20 @@ func (p *partitions) getAndCacheActivePartition(rootChecker func(StatCommander,
 
        activePartition, err := getRootFromMountedDevices(p, rootChecker, mountedDevices, rootDevice)
        if err != nil {
+               // If we reach this point, we have not been able to find a match
+               // based on mounted device.
+               //
+               // Fall-back to configuration and environment only!
+               if checkBootEnvAndRootPartitionMatch(bootEnvBootPart, p.rootfsPartA) {
+                       p.active = p.rootfsPartA
+                       log.Debug("Setting active partition from configuration and environment: ", p.active)
+                       return p.active, nil
+               }
+               if checkBootEnvAndRootPartitionMatch(bootEnvBootPart, p.rootfsPartB) {
+                       p.active = p.rootfsPartB
+                       log.Debug("Setting active partition from configuration and environment: ", p.active)
+                       return p.active, nil
+               }
                return "", err
        }
        if checkBootEnvAndRootPartitionMatch(bootEnvBootPart, activePartition) {
```

[1] `getMountedDevices()` is really incorrectly named. It should be called `getDeviceFiles()`

Reply:
> getMountedDevices() gets the device files on the system, and then try to match this against the major and minor number of the / descriptor. Does squashfs not have a device file to match against?

No, squashfs nor UBI can match major/minor numbers of / descriptor.  I did a fix for UBI when that was integrated where it fall-backed to the `mount` output and comparing that to devices. In this case `mount` does not give any hints of device used and hence that it does not work.

This is a special case, where UBI volumes are used, with squashfs content on top of ubiblocks (read-only MTD blocks on top of UBI). 

I like your approach more and will try that out. I was a bit worried that my approach removes the error paths.

Thank you @kacf